ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestCase.java

```diff
@@ -295,4 +314,208 @@ private static String standardizeTopology(final String topology) {
+      final SchemaRegistryClient schemaRegistryClient) {
+    final Map<String, List<FakeKafkaRecord>> expectedOutput = getExpectedRecordsMap(
+        schemaRegistryClient);
+    final Map<String, List<FakeKafkaRecord>> outputRecordsFromKafka = new HashMap<>();
+    expectedOutput.keySet().forEach(
+        kafkaTopicName -> outputRecordsFromKafka.put(
+            kafkaTopicName,
+            fakeKafkaService.readRecordsFromTopic(kafkaTopicName)
+        )
+    );
+    expectedOutput.keySet().forEach(kafkaTopic -> validateTopicData(
+        expectedOutput.get(kafkaTopic),
+        outputRecordsFromKafka.get(kafkaTopic),
+        schemaRegistryClient
+    ));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static void validateTopicData(
+      final List<FakeKafkaRecord> expected,
+      final List<FakeKafkaRecord> actual,
+      final SchemaRegistryClient schemaRegistryClient) {
+    assertThat(expected.size(), equalTo(actual.size()));
+    for (int i = 0; i < expected.size(); i++) {
+      final ProducerRecord actualProducerRecord = actual.get(i).getProducerRecord();
+      final ProducerRecord expectedProducerRecord = expected.get(i).getProducerRecord();
+      final Object value;
+      if (expected.get(i).getTestRecord().topic.getSerdeSupplier()
+          instanceof ValueSpecJsonSerdeSupplier) {
+        value = new String((byte[]) expectedProducerRecord.value(), Charset.forName("UTF-8"));
+      } else {
+        final Deserializer deserializer = expected.get(i).getTestRecord().topic
+            .getDeserializer(schemaRegistryClient);
+        value = deserializer.deserialize("", (byte[]) expectedProducerRecord.value());
+      }
+
+      if (
+          !actualProducerRecord.timestamp().equals(expectedProducerRecord.timestamp())
+          || !actualProducerRecord.key().equals(expectedProducerRecord.key())
+          || !actualProducerRecord.value().equals(value)
+          ) {
+        final AssertionError error = new AssertionError(
+            "Expected <" + expectedProducerRecord.key() + ", "
+                + value + "> with timestamp="
+                + expectedProducerRecord.timestamp()
+                + " but was <" + actualProducerRecord.key() + ", "
+                + actualProducerRecord.value() + "> with timestamp="
+                + actualProducerRecord.timestamp());
+        error.printStackTrace();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<String, List<FakeKafkaRecord>> getExpectedRecordsMap(
+      final SchemaRegistryClient schemaRegistryClient) {
+    final Map<String, List<FakeKafkaRecord>> outputRecordsFromTest = new HashMap<>();
+    outputRecords.forEach(
+        record ->  {
+          if (!outputRecordsFromTest.containsKey(record.topic.getName())) {
+            outputRecordsFromTest.put(record.topic.getName(), new ArrayList<>());
+          }
+          outputRecordsFromTest.get(record.topic.getName()).add(
+              new FakeKafkaRecord(record, new ProducerRecord(
```

Comment:
similar comment, what about having a `FakeKafakRecord#of(Record, byte[] bytes)` and moving some of this boilerplate to a simpler class?